router/src/panic.c

```diff
@@ -0,0 +1,332 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+//
+// Code for unwinding the stack and printing debug information when a non-recoverable signal occurs.
+//
+
+#define _GNU_SOURCE  // to get gettid()
+
+#include "config.h"
+
+#include "qpid/dispatch/threading.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <signal.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#ifdef HAVE_LIBUNWIND
+#define UNW_LOCAL_ONLY
+#include <libunwind.h>
+#endif
+
+static void panic_signal_handler(int signum, siginfo_t *siginfo, void *ucontext);
+
+// Define those signals which will be handled.
+//
+typedef struct {
+    int         signal;
+    const char *name;
+} panic_signal_info_t;
+
+// clang-format off
+static const panic_signal_info_t panic_signals[] = {
+    { .signal = SIGABRT, .name = "SIGABRT" },
+    { .signal = SIGBUS,  .name = "SIGBUS" },
+    { .signal = SIGFPE,  .name = "SIGFPE" },
+    { .signal = SIGILL,  .name = "SIGILL" },
+    { .signal = SIGSEGV, .name="SIGSEGV" },
+    {0}
+};
+// clang-format on
+
+/**
+ * Install the panic signal handler. This is done early in the router initialization so do not attempt to log or use the
+ * alloc pool, etc. This is not called during a signal handler so there is no need to avoid async signal unsafe calls.
+ */
+void panic_handler_init(void)
+{
+    if (getenv("SKUPPER_ROUTER_DISABLE_PANIC_HANDLER") == 0) {
+        struct sigaction sa = {
+            // use SA_RESETHAND since if the stack unwind fails the default signal handler (coredump) will be invoked
+            .sa_flags     = SA_SIGINFO | SA_RESETHAND,
+            .sa_sigaction = panic_signal_handler,
+        };
+
+        sigemptyset(&sa.sa_mask);
+
+        for (int i = 0; panic_signals[i].signal != 0; ++i) {
+            sigaction(panic_signals[i].signal, &sa, 0);
+        }
+    }
+}
+
+//
+// The remaining routines are invoked by a signal handler. They must not invoke any function that is not Async Signal
+// Safe. See man signal(7) and man signal-safety(7). Ignore this sage advice at your on peril..
+//
+
+#define BACKTRACE_LIMIT 64
+#define BUFFER_SIZE     256
+static unsigned char buffer[BUFFER_SIZE];
+
+// expected to be defined by the linker
+//
+extern char  __executable_start[];
+extern char  __etext[];
+extern char *program_invocation_name;
+
+// async signal safe
+//
+static void print(const char *str)
+{
+    size_t ignore = write(STDERR_FILENO, str, strlen((char *) str));
+    (void) ignore;
+    fsync(STDERR_FILENO);
```

Comment:
stderr by default does not buffer, but I guess there is no harm in calling fsync, that is on the safe list